File: scripts/gen-swatch.mjs
import { readFileSync, writeFileSync } from "fs";
import { resolve, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = resolve(__dirname, "..");
const BRAND_DIR = resolve(ROOT, "assets", "brand");

// Load .env
const envText = readFileSync(resolve(ROOT, ".env"), "utf8");
for (const line of envText.split("\n")) {
  const trimmed = line.trim();
  if (!trimmed || trimmed.startsWith("#")) continue;
  const eq = trimmed.indexOf("=");
  if (eq > 0) process.env[trimmed.slice(0, eq).trim()] = trimmed.slice(eq + 1).trim();
}

// Focused color swatch prompt with text-accuracy techniques
const COLOR_SWATCH_PROMPT = [
  'Intended use: professional brand color palette swatch for a design system document.',
  '',
  'Layout: A clean grid of color swatches on a dark background, color #08090F.',
  'Each swatch is a large rounded rectangle filled with the exact color.',
  'Below each swatch, the hex code is printed in a small sans-serif font.',
  '',
  'Row 1 — Backgrounds (3 swatches):',
  '  Swatch 1: fill #08090F, label "#08090F" and "Base"',
  '  Swatch 2: fill #0F1219, label "#0F1219" and "Card"',
  '  Swatch 3: fill #1A1F2E, label "#1A1F2E" and "Elevated"',
  '',
  'Row 2 — Accents (4 swatches):',
  '  Swatch 4: fill #667EEA, label "#667EEA" and "Indigo"',
  '  Swatch 5: fill #764BA2, label "#764BA2" and "Purple"',
  '  Swatch 6: fill #F093FB, label "#F093FB" and "Pink"',
  '  Swatch 7: fill #56B8DE, label "#56B8DE" and "Frost Cyan"',
  '',
  'Row 3 — Text colors (3 swatches):',
  '  Swatch 8: fill #FFFFFF, label "#FFFFFF" and "Bright"',
  '  Swatch 9: fill #C8D0E0, label "#C8D0E0" and "Default"',
  '  Swatch 10: fill #6B7A99, label "#6B7A99" and "Muted"',
  '',
  'Row 4 — Status colors (4 swatches):',
  '  Swatch 11: fill #00E5A0, label "#00E5A0" and "Success"',
  '  Swatch 12: fill #FBBF24, label "#FBBF24" and "Warning"',
  '  Swatch 13: fill #FB7185, label "#FB7185" and "Danger"',
  '  Swatch 14: fill #38BDF8, label "#38BDF8" and "Info"',
  '',
  'Below all swatches, show one wide horizontal gradient bar from #667EEA through #764BA2 to #F093FB, labeled "Brand Gradient".',
  '',
  'Constraints:',
  '- Each color swatch must be exactly the hex color specified — no approximation.',
  '- All text must be crisp, legible, and correctly spelled.',
  '- Flat design only. No shadows, no 3D, no reflections.',
  '- Dark swatches (#08090F, #0F1219, #1A1F2E) should have a thin white/gray border so they are visible against the dark background.',
  '- Generous spacing between swatches. Clean, minimal, professional.',
].join('\n');

async function genOpenAI() {
  console.log("Generating with GPT Image 1.5...");
  const key = process.env.OPENAI_API_KEY;
  if (!key) { console.log("ERROR: OPENAI_API_KEY not set"); return; }

  const resp = await fetch("https://api.openai.com/v1/images/generations", {
    method: "POST",
    headers: { "Content-Type": "application/json", Authorization: `Bearer ${key}` },
    body: JSON.stringify({
      model: "gpt-image-1",
      prompt: COLOR_SWATCH_PROMPT,
      n: 1,
      size: "1536x1024",
      quality: "high",
    }),
  });
  const data = await resp.json();
  if (data.error) { console.log("OpenAI error:", data.error.message); return; }

  // gpt-image-1 returns b64_json by default
  if (data.data[0].b64_json) {
    const buf = Buffer.from(data.data[0].b64_json, "base64");
    const out = resolve(BRAND_DIR, "frost-color-swatch-openai.png");
    writeFileSync(out, buf);
    console.log(`Saved OpenAI image: ${out} (${buf.length} bytes)`);
  } else if (data.data[0].url) {
    const imgResp = await fetch(data.data[0].url);
    const buf = Buffer.from(await imgResp.arrayBuffer());
    const out = resolve(BRAND_DIR, "frost-color-swatch-openai.png");
    writeFileSync(out, buf);
    console.log(`Saved OpenAI image: ${out} (${buf.length} bytes)`);
  }
}

async function genGemini() {
  console.log("Generating with Gemini...");
  const key = process.env.GEMINI_API_KEY;
  if (!key) { console.log("ERROR: GEMINI_API_KEY not set"); return; }

  const resp = await fetch(
    `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash-exp-image-generation:generateContent?key=${key}`,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        contents: [{ parts: [{ text: "Generate an image:\n\n" + COLOR_SWATCH_PROMPT }] }],
        generationConfig: { responseModalities: ["TEXT", "IMAGE"] },
      }),
    }
  );
  const data = await resp.json();
  if (data.error) { console.log("Gemini error:", data.error.message); return; }

  const parts = data.candidates?.[0]?.content?.parts || [];
  let saved = false;
  for (const part of parts) {
    if (part.inlineData) {
      const buf = Buffer.from(part.inlineData.data, "base64");
      const out = resolve(BRAND_DIR, "frost-color-swatch-gemini.png");
      writeFileSync(out, buf);
      console.log(`Saved Gemini image: ${out} (${buf.length} bytes)`);
      saved = true;
    } else if (part.text) {
      console.log("Gemini text:", part.text.slice(0, 200));
    }
  }
  if (!saved) {
    console.log("No image in Gemini response");
    console.log(JSON.stringify(data).slice(0, 500));
  }
}

async function genIdeogram() {
  console.log("Generating with Ideogram 3...");
  const key = process.env.IDEOGRAM_API_KEY;
  if (!key) { console.log("ERROR: IDEOGRAM_API_KEY not set"); return; }

  const resp = await fetch("https://api.ideogram.ai/generate", {
    method: "POST",
    headers: {
      "Api-Key": key,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      image_request: {
        prompt: COLOR_SWATCH_PROMPT,
        model: "V_2",
        magic_prompt_option: "OFF",
        aspect_ratio: "ASPECT_3_2",
        style_type: "DESIGN",
      },
    }),
  });
  const data = await resp.json();
  if (data.error || !data.data) {
    console.log("Ideogram error:", JSON.stringify(data).slice(0, 500));
    return;
  }

  const url = data.data[0]?.url;
  if (url) {
    const imgResp = await fetch(url);
    const buf = Buffer.from(await imgResp.arrayBuffer());
    const out = resolve(BRAND_DIR, "frost-color-swatch-ideogram.png");
    writeFileSync(out, buf);
    console.log(`Saved Ideogram image: ${out} (${buf.length} bytes)`);
  }
}

const GENERATORS = { openai: genOpenAI, gemini: genGemini, ideogram: genIdeogram };
const target = process.argv[2] || "all";
const tasks = [];

if (target === "all") {
  tasks.push(...Object.values(GENERATORS).map(fn => fn()));
} else {
  for (const name of target.split(",")) {
    const fn = GENERATORS[name.trim()];
    if (fn) tasks.push(fn());
    else console.log(`Unknown generator: ${name}. Options: ${Object.keys(GENERATORS).join(", ")}`);
  }
}

await Promise.all(tasks);

File: scripts/gen-swatch.py
"""Generate Frost brand swatch using DALL-E and Gemini APIs."""
import os, json, base64, urllib.request, sys

# Load env
from pathlib import Path
env_path = Path(__file__).resolve().parent.parent / ".env"
for line in env_path.read_text().splitlines():
    line = line.strip()
    if line and not line.startswith("#") and "=" in line:
        k, v = line.split("=", 1)
        os.environ[k.strip()] = v.strip()

PROMPT = (
    "Brand identity swatch board for 'Frost', a dark-themed AI-powered cannabis operations platform. "
    "Clean modern brand board on near-black background (#08090F). Organized in a grid with clear sections. "
    "TOP LEFT: Bold geometric six-pointed snowflake icon in sky-blue (#56B8DE) next to the word FROST in "
    "wide-tracked uppercase sans-serif, also sky-blue. Show both the full wordmark and the standalone snowflake icon. "
    "TOP RIGHT: Color palette as large rounded-rectangle swatches with hex codes beneath each: "
    "#08090F (Base Black), #0F1219 (Card Dark), #1A1F2E (Elevated), #667EEA (Accent Indigo), "
    "#764BA2 (Accent Purple), #F093FB (Accent Pink), #56B8DE (Logo Cyan), #C8D0E0 (Body Silver), "
    "#00E5A0 (Success Green), #FBBF24 (Warning Amber), #FB7185 (Danger Rose). "
    "Also show the brand gradient as a wide horizontal bar: from #667EEA through #764BA2 to #F093FB. "
    "BOTTOM LEFT: Typography specimens. Heading: 'Frost Operations' in bold white, large sans-serif. "
    "Body: 'AI-powered cannabis platform with long-term RAG memory' in silver (#C8D0E0), medium sans-serif. "
    "Code: 'queryFn: fetchInventory()' in muted blue-gray (#6B7A99), small monospace. "
    "BOTTOM RIGHT: A small UI card mockup on #0F1219 background with rounded corners, thin #1A1F2E border, "
    "indigo (#667EEA) accent bar at top, white title text, silver body text, small green (#00E5A0) status badge. "
    "Style: ultra-clean, minimal, professional brand board. Flat design, no 3D. Dark mode throughout. "
    "Generous whitespace. Cold, precise, premium aesthetic."
)

BRAND_DIR = Path(__file__).resolve().parent.parent / "assets" / "brand"

def gen_dalle():
    print("Generating with DALL-E 3...")
    key = os.environ.get("OPENAI_API_KEY")
    if not key:
        print("ERROR: OPENAI_API_KEY not set"); return
    body = json.dumps({
        "model": "dall-e-3",
        "prompt": PROMPT,
        "n": 1,
        "size": "1792x1024",
        "quality": "hd"
    }).encode()
    req = urllib.request.Request(
        "https://api.openai.com/v1/images/generations",
        data=body,
        headers={"Content-Type": "application/json", "Authorization": f"Bearer {key}"}
    )
    try:
        resp = urllib.request.urlopen(req, timeout=120)
        data = json.loads(resp.read())
        url = data["data"][0]["url"]
        img = urllib.request.urlopen(url, timeout=60).read()
        out = BRAND_DIR / "frost-brand-swatch-dalle.png"
        out.write_bytes(img)
        print(f"Saved DALL-E image: {out} ({len(img)} bytes)")
    except Exception as e:
        print(f"DALL-E error: {e}")
        if hasattr(e, 'read'):
            print(e.read().decode())

def gen_gemini():
    print("Generating with Gemini Imagen...")
    key = os.environ.get("GEMINI_API_KEY")
    if not key:
        print("ERROR: GEMINI_API_KEY not set"); return
    body = json.dumps({
        "contents": [{"parts": [{"text": "Generate an image: " + PROMPT}]}],
        "generationConfig": {"responseModalities": ["TEXT", "IMAGE"]}
    }).encode()
    req = urllib.request.Request(
        f"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.0-flash-exp:generateContent?key={key}",
        data=body,
        headers={"Content-Type": "application/json"}
    )
    try:
        resp = urllib.request.urlopen(req, timeout=120)
        data = json.loads(resp.read())
        parts = data.get("candidates", [{}])[0].get("content", {}).get("parts", [])
        saved = False
        for part in parts:
            if "inlineData" in part:
                img = base64.b64decode(part["inlineData"]["data"])
                out = BRAND_DIR / "frost-brand-swatch-gemini.png"
                out.write_bytes(img)
                print(f"Saved Gemini image: {out} ({len(img)} bytes)")
                saved = True
            elif "text" in part:
                print(f"Gemini text: {part['text'][:200]}")
        if not saved:
            print("No image in Gemini response")
            print(json.dumps(data, indent=2)[:500])
    except Exception as e:
        print(f"Gemini error: {e}")
        if hasattr(e, 'read'):
            print(e.read().decode())

if __name__ == "__main__":
    target = sys.argv[1] if len(sys.argv) > 1 else "both"
    if target in ("dalle", "both"):
        gen_dalle()
    if target in ("gemini", "both"):
        gen_gemini()
